### PR TITLE
[WIP] Upgrade to xunit 2.1

### PR DESCRIPTION
**Pull request by damianh**
- Needed for improved and better async support
 - Will carry forward into dnx world later

(Will Travis accept it ?? :)

### CODE REVIEW

File: src/Nancy.Tests/Unit/Diagnostics/DefaultRequestTraceFactoryFixture.cs [hunk 1]
```diff
@@ -62,10 +62,7 @@ public void Should_initialize_case_insensitive_items_collection_when_case_sensit
                 trace.Items.Add("foo", "bar");
 
                 // Then
-                Assert.DoesNotThrow(() =>
-                {
-                    var value = trace.Items["FOO"];
-                });
+                var value = trace.Items["FOO"];
```
Comment (damianh) [inline src/Nancy.Tests/Unit/Diagnostics/DefaultRequestTraceFactoryFixture.cs]: @NancyFx/most-valued-minions please have a read of this https://github.com/xunit/xunit/issues/188 . Are we happy with leaving this as is, or would you prefer we used `Record.Exception` as suggested here https://github.com/xunit/xunit/issues/188#issuecomment-86584802 ?

**Comment by phillip-haydon** [comment]
Could just write an extension method to implement the same API. 

I don't really agree with why it was removed. But we either keep the same API (via our own extension method on Assert) or change all usages to use the `Record.Exception` and `Assert.Null(...)`

**Comment by phillip-haydon** [comment]
An extension method may mean less code changes and we can progressively re-do all the usages over time.

**Comment by damianh** [comment]
We're kinda working against the philosophy of xunit here and it's only used in 7 places...

**Comment by damianh** [comment]
Ftr, I agree with Brad on this.

File: src/Nancy.Tests/Unit/Diagnostics/DefaultRequestTraceFactoryFixture.cs (hunk 1 above)
Comment (grumpydev) [inline src/Nancy.Tests/Unit/Diagnostics/DefaultRequestTraceFactoryFixture.cs]: Record.Exception gets my vote.. hate the Assert.Throws and Assert.DoesNotThrow stuff.

**Comment by phillip-haydon** [comment]
If it's 7 places then just convert it to the new style.

File: src/Nancy.Tests/Unit/Diagnostics/DefaultRequestTraceFactoryFixture.cs (hunk 1 above)
Comment (damianh) [inline src/Nancy.Tests/Unit/Diagnostics/DefaultRequestTraceFactoryFixture.cs]: Record.Exception(() => trace.Items["FOO"])
        .ShouldBeNull();

^ Look reasonable?

Comment (jchannon) [inline src/Nancy.Tests/Unit/Diagnostics/DefaultRequestTraceFactoryFixture.cs]: Isn't the point of using Record to then assert the type of the expected
exception?

**Comment by damianh** [comment]
Xunit 2.x brings with it parallel testing by default, noticed a bunch of failing tests that then subsequently passed when run individually. FakeItEasy has problems with parallel test execution. https://github.com/FakeItEasy/FakeItEasy/issues/60 . On company projected we removed FIE. (I tend to do acceptance tests anyway). Think I'll disable parallel test execution in each assembly for now.

**Comment by damianh** [comment]
`D, [2015-11-01T19:51:28.019649 #28204] DEBUG -- : Executing XUnit: "/home/travis/build/NancyFx/Nancy/tools/xunit/xunitmono.sh" src/Nancy.Authentication.Basic.Tests/bin/MonoRelease/Nancy.Authentication.Basic.Tests.dll 
System.TypeLoadException: Could not load type 'Xunit.Sdk.XunitTestFrameworkDiscoverer, xunit.execution.desktop, Version=2.1.0.3179, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c'.`

And there we are :|

**Comment by damianh** [comment]
:white_check_mark: All tests via ncrunch
:white_check_mark: All tests via resharper
:white_check_mark: All tests via rake on windows

Both Travis and TeamCity are probably having a problem .net versions supported / installed.

**Comment by damianh** [comment]
`[19:33:28][Execute compile]   C:\Windows\Microsoft.NET\Framework\v4.0.30319\Csc.exe /noconfig /nowarn:1701,1702 /nostdlib+ /errorreport:prompt /warn:4 /define:TRACE /highentropyva+ .... "C:\TeamCity\buildAgent\temp\buildTmp\.NETFramework,Version=v4.5.AssemblyAttributes.cs"
[19:33:30][Execute compile] xUnitExtensions\SkippableFactAttribute.cs(18,18): error CS0012: The type 'System.Attribute' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. [C:\TeamCity\buildAgent\work\18e602d45da8b563\src\Nancy.Tests\Nancy.Tests.csproj]
[19:33:30][Execute compile] Helpers\ExceptionExtensionsFixture.cs(9,10): error CS0616: 'Xunit.FactAttribute' is not an attribute class [C:\TeamCity\buildAgent\work\18e602d45da8b563\src\Nancy.Tests\Nancy.Tests.csproj]
[19:33:30][Execute compile] Helpers\ExceptionExtensionsFixture.cs(18,10): error CS0616: 'Xunit.FactAttribute' is not an attribute class [C:\TeamCity\buildAgent\work\18e602d45da8b563\src\Nancy.Tests\Nancy.Tests.csproj]`

File: src/Nancy.Authentication.Basic/BasicAuthentication.cs [hunk 2]
```diff
@@ -2,6 +2,7 @@
 {
     using System;
     using System.Text;
+
```
Comment (khellang) [inline src/Nancy.Authentication.Basic/BasicAuthentication.cs]: What's up with all these?

**Comment by damianh** [comment]
Not going to be able to fix this until https://github.com/xunit/xunit/issues/158 is resolved.

**Comment by damianh** [comment]
Rebase resulted in a mess. This should be re-done at the appropriate time.

File: src/Nancy.Authentication.Basic/BasicAuthentication.cs (hunk 2 above)
Comment (damianh) [inline src/Nancy.Authentication.Basic/BasicAuthentication.cs]: PR is fucked. Skip.

**Comment by adamralph** [comment]
FYI the reports we are getting from the wild is that the FakeItEasy 2.0 betas are performing very well with parallel tests in xunit 2.1. The only thing that continues to be a thread-unsafe pain is [`FakeScope`](https://github.com/FakeItEasy/FakeItEasy/issues/484), but I see you are only using that in [one place](https://github.com/NancyFx/Nancy/blob/2cf0a22ea83c13c3e10fa8ef784810c3dc05f079/src/Nancy.Tests/Unit/Bootstrapper/NancyBootstrapperBaseFixture.cs#L134-L144).

**Pull request closed without merging**